lib/Subscriber.js

```diff
@@ -0,0 +1,592 @@
+        for (const subscribe of this._queue)
+        {
+          logger.debug('dequeue subscribe');
+
+          this._sendSubsequentSubscribe(subscribe.body, subscribe.headers);
+        }
+      }
+
+      // Check expires value.
+      let expires_value = response.getHeader('expires');
+
+      if (expires_value !== 0 && !expires_value)
+      {
+        logger.warn('response without Expires header');
+
+        // RFC 6665 3.1.1 subscribe OK response must contain Expires header.
+        // Use workaround expires value.
+        expires_value = '900';
+      }
+
+      const expires = parseInt(expires_value);
+
+      if (expires > 0)
+      {
+        this._scheduleSubscribe(expires);
+      }
+    } 
+    else if (response.status_code === 401 || response.status_code === 407)
+    {
+      this._dialogTerminated(C.SUBSCRIBE_FAILED_AUTHENTICATION);
+    }
+    else if (response.status_code >= 300) 
+    {
+      this._dialogTerminated(C.SUBSCRIBE_NON_OK_RESPONSE);
+    }
+  }
+
+  _sendSubsequentSubscribe(body, headers)
+  {
+    if (this._state === C.STATE_TERMINATED)
+    {
+      return;  
+    }
+    
+    if (!this._dialog)
+    {
+      logger.debug('enqueue subscribe');
+
+      this._queue.push({ body, headers: headers.slice() });
+
+      return;
+    }
+    
+    if (body) 
+    {
+      if (!this._content_type)
+      { 
+        throw new TypeError('content_type is undefined');
+      }
+
+      headers = headers.slice();
+      headers.push(`Content-Type: ${this._content_type}`);
+    }
+
+    this._dialog.sendRequest(JsSIP_C.SUBSCRIBE, { 
+      body, 
+      extraHeaders  : headers,
+      eventHandlers : {
+        onRequestTimeout : () =>
+        {
+          this.onRequestTimeout();
+        },
+        onTransportError : () =>
+        {
+          this.onTransportError();
+        },
+        onSuccessResponse : (response) =>
+        {
+          this._receiveSubscribeResponse(response);
+        }, 
+        onErrorResponse : (response) =>
+        {
+          this._receiveSubscribeResponse(response);
+        }, 
+        onDialogError : (response) =>
+        {
+          this._receiveSubscribeResponse(response);
+        } 
+      }
+    });
+  }
+
+  _dialogTerminated(terminationCode, reason = undefined, retryAfter = undefined) 
+  {
+    // To prevent duplicate emit terminated event.
+    if (this._state === C.STATE_TERMINATED)
+    {
+      return;
+    }
+
+    this._state = C.STATE_TERMINATED;
+
+    // Clear timers.
+    clearTimeout(this._expires_timer);
+    clearTimeout(this._unsubscribe_timeout_timer);
+
+    if (this._dialog)
+    {
+      this._dialog.terminate();
+      this._dialog = null;
+    }
+
+    logger.debug(`emit "terminated" code=${terminationCode}`);
+    this.emit('terminated', terminationCode, reason, retryAfter);
+  }
+
+  _scheduleSubscribe(expires) 
+  {
+    const timeout = expires >= 140 ? (expires * 1000 / 2) 
```

Comment:
Please, add the corresponding log here indicating the rationale of this calculus.

Reply:
/*
      If expires less than 140 seconds we don't support Chrome high throttling timer mode. In the case re-SUBSCRIBE will send 5 seconds before expiration.
	  
	  Chrome high throttling mode: in the worst case timer fired 60 seconds later.
	  We give the server 10 seconds to make sure it will execute the command even if it is heavily loaded. So we order re-subscription time not later than 70 seconds before expiration. We also add randomization.
	  So time calculated as half time interval + (half interval - 70) * random.
	  
	  E.g. if expires is 140, re-subscribe will be ordered in 70 seconds.
	       if expires is 600, re-subscribe will be ordered in 300 + (0 .. 230) seconds.
	 */
  
I feel that explanation is too big to comment ;-)